### ISSUE TITLE
The behavior of numpy.apply_along_axis function is not what expected.

### ISSUE THREAD

**Issue by yorklee70** [numpy/numpy#5248]
When I use python set as elelments of a numpy array, the behavior of numpy.apply_along_axis function is not what expected. For example, I want to get the union set for each column using the following codes:

```python
import numpy
import operator
d = numpy.array([  [ set([1, 11]), set([2, 22]), set([3, 33]) ],
                              [ set([4, 44]), set([5, 55]), set([6, 66]) ]  ])
numpy.apply_along_axis( lambda a:reduce(operator.**or**, a), 0, d)
```
i.e. I expect to get a result like:

        array([ set(1, 11, 4, 44), set(2, 22, 5, 55), set(3, 33, 6, 66) ]

however,  the result is:

        array([ [ set(1, 11, 4, 44), set(2, 22, 5, 55), set(3, 33, 6, 66) ],
                   [ set(1, 11, 4, 44), set(2, 22, 5, 55), set(3, 33, 6, 66) ],
                   [ set(1, 11, 4, 44), set(2, 22, 5, 55), set(3, 33, 6, 66) ],
                   [ set(1, 11, 4, 44), set(2, 22, 5, 55), set(3, 33, 6, 66) ] ])

The expected result was repeated 4 times, which is determined by the first element, actually it equals to the length of the first element.

Can this be viewed as a bug? Anyway it is different with the way of the "apply" function in R language.

York

**Comment by eric-wieser** [comment]
So, this was kind of fixed by #8441, but also kind of not:

```
>>> import functools
>>> import operator
>>> import numpy as np
>>> d = np.array([  [ set([1, 11]), set([2, 22]), set([3, 33]) ],  [ set([4, 44]), set([5, 55]), set([6, 66]) ]  ])
>>> np.apply_along_axis( lambda a: functools.reduce(operator.or_, a), 0, d)
array([array({1, 11, 4, 44}, dtype=object),
       array({2, 5, 22, 55}, dtype=object),
       array({33, 66, 3, 6}, dtype=object)], dtype=object)
```

**Comment by eric-wieser** [comment]
Closing in favor of #8642
